### PR TITLE
[receiver/haproxy] add support for http

### LINKED ISSUE #24440: Haproxyreceiver HTTP not supported
### Component(s)

receiver/haproxy

## Description
HAProxy receiver readme mentions support for HTTP, but it does not work for me.

Seems to me like the functionality support is planned but not implemented yet.
If that is the case, I would be ok with that state being mentioned in the readme.

## Steps to Reproduce

1. Run v81 collector docker image
2. Run haproxy:2.8.1 docker image (and expose HTTP stats endpoint by configuring haproxy.cfg)
```
frontend stats
        mode http
        bind *:8404
        stats enable
        stats uri /stats
```
3. Try to scrape the HTTP endpoint with the receiver. (test in browser that it really is exposed)
```
receivers:
  haproxy:
    endpoint: "http://{haproxy-container-ip}:8408/stats"
    collection_interval: 30s
```

## Expected Result
HTTP endpoint scraped as mentioned in readme.
Or readme saying it does not work for HTTP.

## Actual Result
HTTP endpoint not scraped, looks like it uses the same code path as for the socket, resulting in a fail.

### Collector version

0.81.0

## Environment
Collector docker image: otel/opentelemetry-collector-contrib:0.81.0

HaProxy Dockerfile
```
FROM haproxy:2.8.1 as updated
COPY haproxy.cfg /usr/local/etc/haproxy/haproxy.cfg
```

I just added this to haproxy.cfg to enable stats on 8404/stats 
```
frontend stats
        mode http
        bind *:8404
        stats enable
        stats uri /stats
```
That exposed this.
image

### OpenTelemetry Collector configuration

```yaml
receivers:
  haproxy:
    endpoint: "http://{haproxy-container-ip}:8404/stats"
    collection_interval: 30s

exporters:
  otlp:
    endpoint: $OTEL_ENDPOINT
    tls:
      insecure: false
    headers: {"Authorization": $OTEL_TOKEN}

service:
  pipelines:
    metrics:
      receivers: [haproxy]
      exporters: [otlp]
```

### Log output

```shell
2023-07-24 09:22:00 2023-07-24T07:22:00.979Z    error   scraperhelper/scrapercontroller.go:213  Error scraping metrics  {"kind": "receiver", "name": "haproxy", "data_type": "metrics", "error": "dial unix http://localhost:8404/stats: connect: no such file or directory", "scraper": ""}
2023-07-24 09:22:00 go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport
2023-07-24 09:22:00     go.opentelemetry.io/collector/receiver@v0.81.0/scraperhelper/scrapercontroller.go:213
2023-07-24 09:22:00 go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1
2023-07-24 09:22:00     go.opentelemetry.io/collector/receiver@v0.81.0/scraperhelper/scrapercontroller.go:192
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
Support HTTP endpoints as well for haproxy receiver.

**Link to tracking Issue:**
Fixes #24440 

**Testing:**
Integration test

**Documentation:**
N/A

### CODE REVIEW

**Comment by atoulme** [comment]
@akumetsu183  PTAL?

File: receiver/haproxyreceiver/integration_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	"context"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
```
[the remark this answers is not in the document]
Comment (atoulme) [reply receiver/haproxyreceiver/integration_test.go:17]: I have not used it before, I will take a look, thanks for the suggestion.

File: receiver/haproxyreceiver/scraper.go [hunk 2]
```diff
@@ -28,30 +32,60 @@ var (
...
+		if err != nil {
+			return pmetric.NewMetrics(), err
+		}
+	} else {
+		var d net.Dialer
+		c, err := d.DialContext(ctx, "unix", s.cfg.Endpoint)
+		if err != nil {
+			return pmetric.NewMetrics(), err
+		}
+		defer func(c net.Conn) {
```
Comment (fatsheep9146) [inline receiver/haproxyreceiver/scraper.go:66]: I wonder what's the different between this and `defer c.Close()`, is this func is used on purpose?

File: receiver/haproxyreceiver/integration_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build integration
+// +build integration
+
+package haproxyreceiver
+
+import (
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/scraperinttest"
```
Comment (fatsheep9146) [inline receiver/haproxyreceiver/integration_test.go:11]: the format of this import list may not pass lint

File: receiver/haproxyreceiver/scraper.go (hunk 2 above)
Comment (atoulme) [reply receiver/haproxyreceiver/scraper.go:66]: this is just to make explicit that we're not going to care for the error returned by Close() by assigning it to `_`, which removes a warning from IDEs.

File: receiver/haproxyreceiver/scraper.go [hunk 4]
```diff
@@ -28,30 +32,60 @@ var (
...
-	}
-	defer func(c net.Conn) {
-		_ = c.Close()
-	}(c)
-	records, err := s.readStats(c)
-	if err != nil {
-		return pmetric.NewMetrics(), err
+
+	var records []map[string]string
+	if u, notURLerr := url.Parse(s.cfg.Endpoint); notURLerr == nil && (u.Scheme == "http" || u.Scheme == "https") {
```
Comment (MovieStoreGuy) [inline receiver/haproxyreceiver/scraper.go:45]: If you can, I suggest doing !strings.HasPrefix(u.Scheme, "http")` instead ?

**Review by MovieStoreGuy — APPROVED** [review]
One thing to simplify validation logic, but otherwise, 👍🏽

File: receiver/haproxyreceiver/scraper.go (hunk 4 above)
Comment (atoulme) [reply receiver/haproxyreceiver/scraper.go:45]:
```suggestion
	if u, notURLerr := url.Parse(s.cfg.Endpoint); notURLerr == nil && !strings.HasPrefix(u.Scheme, "http") {
```

File: receiver/haproxyreceiver/scraper.go [hunk 5]
```diff
@@ -28,30 +32,60 @@ var (
...
-	}
-	defer func(c net.Conn) {
-		_ = c.Close()
-	}(c)
-	records, err := s.readStats(c)
-	if err != nil {
-		return pmetric.NewMetrics(), err
+
+	var records []map[string]string
+	if u, notURLerr := url.Parse(s.cfg.Endpoint); notURLerr == nil && !strings.HasPrefix(u.Scheme, "http") {
```
Comment (atoulme) [inline receiver/haproxyreceiver/scraper.go:45]:
```suggestion
	if u, notURLerr := url.Parse(s.cfg.Endpoint); notURLerr == nil && strings.HasPrefix(u.Scheme, "http") {
```

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by atoulme** [comment]
I have no idea why the lint for other group is failing. I tried to address it elsewhere and still run into issues. I filed #25875 to follow up and a PR, but it also fails. Not sure how to fix this, it's clearly unrelated to those code changes.

**Pull request merged**